Filters dev

ekala opened the pull request:
Implements the river level filtering mechanism described in #188. The approach to the filters is the same as that of the channels on both the UX/UI but with slight modifications. The available filters are specified in `application/config/filters.php.template` 

River filters settings page
![River filter settings](https://dl.dropbox.com/u/2635815/river-filter-settings.png)

Activating a filter for use
![Add Filter](https://dl.dropbox.com/u/2635815/add-filter.png)

Adding a filter parameter
![Add filter parameter](https://dl.dropbox.com/u/2635815/add-filter-parameter.png)

ekala committed:
Implements river filters - #188

* The approach for the river filters is very much like that of the
river channels - use of a config (config/filters.template.php) file to
define the structure of the filter. As with channels and channel filter
options, each filter has a set of filter parameters. This therefore
introduces two new tables; river_filters and river_filter_parameters.

* I've also relaxed CSRF token regeneration. The tokens are now tied to
a user session. Once the session expires, the token also expires. This
now allows POST requests to be processed from more than one browser tab

* Deleted unused controllers and views from the repository

* TODO: Implement geo-fence filter and introduce configuration
parameter that will allow the site administrator to limit use of the
river filters

ekala committed:
Naming convention update

* Suffix the update script with the name of the affected tables or
relevant subject

ekala committed:
Add "IF NOT EXISTS" clause to river filters DDL

ekala committed:
Purge drops from rivers when filter match fails

* The post processor examines a drop and the river it's destined for vs
its filter. If the drop doesn't match the filter, it's left out of the
river. Since drops only get displayed when processing is complete, we
have to purge them from their respective rivers before their processing
status flag is flipped.

ekala committed:
Return status string in drop API and query modifications

* Drops are only posted via the API once they've undergone
post-processing so there's no need for the REST endpoint to return the
list of freshly created drop entries

* Discard use of "droplet_processed" field (performance impedance
factor) when fetching drops since only processed drops get posted to
the droplets table

69mb commented on application/classes/model/river.php:
```diff
@@ -298,7 +300,6 @@ public static function get_droplets($user_id, $river_id, $drop_id = 0, $page = 1
 			    ->on('droplets.identity_id', '=', 'identities.id')
 				->where('rivers_droplets.droplet_date_pub', '>', '0000-00-00 00:00:00')
 			    ->where('rivers_droplets.river_id', '=', $river_id)
```
Missing ';' at the end of this line.

ekala replied:
Fixed.

69mb commented on themes/default/views/pages/river/settings/layout.php:
```diff
@@ -19,13 +19,16 @@
+					<a href="<?php echo $river_base_url.'/settings/display'; ?>"><?php echo __("Display"); ?></a>
+				</li>
+				<li class="touchcarousel-item <?php if ($active == 'filters' OR ! $active) echo 'active'; ?>">
+					<a href="<?php echo $river_base_url.'/settings/filters'; ?>"><?php echo __("River Filters"); ?></a>
```
Think this should be after the Channels link since they are somewhat related.

ekala committed:
Place filters tab next to the channels tab

* The filters and channels are somewhat related so it makes more sense
to place them next to each other in order to avoid cognitive dissonance.

Recommended by @69mb

ekala committed:
Improve second-level filters

* A named filters facility to allow users to categorize the filters they would like to apply on their river. The named filters are ORd but the individual filter parameters in each named filter are ANDed

* Improved the river creation process. Drops do not appear in a river until the user has completed the river creation i.e. clicks "View my River" in the finish page

* Renamed the river_filter and river_filter_parameter tables to "filters" and "filter_parameters" respectively. This paves the way for bucket-level filters

ekala committed:
Enable/Disable events for channels, filters and rivers

* Run ".enable" or ".disable" events when a channel, filter or river is
enabled or disabled

ekala commented:
@69mb @dkobia I've changed the second-level filters to named filters. The parameters (places, tags, keywords) for each named filter are ANDed while the name filters are OR'd (where a river has more than one named filter)

Screenie for the modified UI

![River Filters](https://dl.dropbox.com/u/2635815/named-filters.png)

dkobia commented on install/sql/swiftriver.sql:
```diff
@@ -783,6 +784,31 @@ CREATE TABLE IF NOT EXISTS `droplet_comments` (
+-- -----------------------------------------------------
+CREATE TABLE IF NOT EXISTS `filters` (
+  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+  `filter_target` varchar(20) NOT NULL DEFAULT COMMENT 'Object on which the filter is to be applied i.e. river or bucket',
```
There is an error on this line ... has **'DEFAULT COMMENT ...'** which is incorrect because COMMENT can't be a DEFAULT.

dkobia replied:
Assuming this was supposed to be **DEFAULT 'river'** ?

ekala committed:
DDL update: Set default filter target to "river"

dkobia commented:
- I seem to be having issues adding a filter parameter. Click and ... nothing. And no console error.
- Also it seems a little confusing to have a filter button, then another 'River Filter' section in the river settings -- wondering how best to separate the two?

ekala commented:
Agreed. I ran out of terminology. Perhaps call them "Saved Filters" or something...

69mb commented:
- I think this can be separated by bringing up a modal to enter filter name when "Add Filter" is clicked instead of in place which is confusing.
- Needs visual feedback: show the loading icon, success + error messages & prevent multiple submission.
- Add 'channel' as a parameter. Thinking I could add rss and sms to my river but I want all SMS and only feed items matching certain keywords. So I would have 2 filters in this river, one with channel rss + my keywords and another with just channel sms to allow all sms drops.
- Also for twitter, I think what we have now as keyword and people channel_filter_options should be filter parameters. Instead of entering keywords on the channel settings page and again on the filters page, we should only select twitter as a channel and it will have no parameters(no channel options in the config file), then on the river filter enter the people and keywords. We could then have a single web.river.new topic that the firehose application will listen to and update itself is a message comes in for a river with a twitter channel. On startup, it will read the filter_parameters table and listen to filter updates.

Take a look at this gist for what I have in mind for the web.river.new message https://gist.github.com/2a4ea18343aadcc6d4a3 maybe this clarifies somewhat?

RSS urls I think are part of channel configuration and can be left in channel_filter_options but the rss scheduler could listen for the new river event and update itself for rivers with an rss channel.

brandonrosage commented:
I’ve completed a solid round of work in the /markup prototype that offers the “Flow” UI within an existing river’s settings. I’d like to bring a similar concept to the river creation process, next, as Emmanuel suggested earlier.

You can find the “Flow” UI in the “ui-dev” branch of the SwiftRiver repo: https://github.com/ushahidi/SwiftRiver/tree/ui-dev

As I discussed before, this approach offers the ability to save filters and distinguish them as either permanent or temporary—but through language on the frontend that’s more consistent with the rest of the app.

SCREENSHOT: http://cl.ly/image/0Q1T2q0C0m2c

To designate a group of filters as temporary, the user effectively makes the group “for their eyes only.” To do this, they select a filter group’s settings and designate “Who can see this group” as “Only you.” Then the filter is applied, but only for that user—it’s not part of the river’s overall makeup.

Take a look and let me know if there are any holes for me to fill.

69mb commented:
Closing this pull request. Superseded by rules.
